staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/celcoststability_test.go

```diff
@@ -0,0 +1,1221 @@
+					},
+				},
+				"listOfMap": []interface{}{
+					map[string]interface{}{
+						"z": "g",
+					},
+				},
+				"listOfObj": []interface{}{
+					map[string]interface{}{
+						"field3": "h",
+					},
+				},
+				"listOfListMap": []interface{}{
+					[]interface{}{
+						map[string]interface{}{
+							"k3": "3",
+							"v3": "i",
+						},
+					},
+				},
+			},
+			schema: objectTypePtr(map[string]schema.Structural{
+				"obj": objectType(map[string]schema.Structural{
+					"field": stringType,
+				}),
+				"mapOfMap": mapType(mapTypePtr(&stringType)),
+				"mapOfObj": mapType(objectTypePtr(map[string]schema.Structural{
+					"field2": stringType,
+				})),
+				"mapOfListMap": mapType(listMapTypePtr([]string{"k"}, objectTypePtr(map[string]schema.Structural{
+					"k": stringType,
+					"v": stringType,
+				}))),
+				"mapOfList": mapType(listTypePtr(&stringType)),
+				"listMapOfObj": listMapType([]string{"k2"}, objectTypePtr(map[string]schema.Structural{
+					"k2": stringType,
+					"v2": stringType,
+				})),
+				"listOfMap": listType(mapTypePtr(&stringType)),
+				"listOfObj": listType(objectTypePtr(map[string]schema.Structural{
+					"field3": stringType,
+				})),
+				"listOfListMap": listType(listMapTypePtr([]string{"k3"}, objectTypePtr(map[string]schema.Structural{
+					"k3": stringType,
+					"v3": stringType,
+				}))),
+			}),
+			expectCost: map[string]uint64{
+				"self.obj.field == 'a'":                                       4,
+				"self.mapOfMap['x']['y'] == 'b'":                              5,
+				"self.mapOfObj['k'].field2 == 'c'":                            5,
+				"self.mapOfListMap['o'].exists(e, e.k == '1' && e.v == 'd')":  14,
+				"self.mapOfList['l'][0] == 'e'":                               5,
+				"self.listMapOfObj.exists(e, e.k2 == '2' && e.v2 == 'f')":     13,
+				"self.listOfMap[0]['z'] == 'g'":                               5,
+				"self.listOfObj[0].field3 == 'h'":                             5,
+				"self.listOfListMap[0].exists(e, e.k3 == '3' && e.v3 == 'i')": 14,
+			},
+		},
+	}
+
+	for _, tt := range cases {
+		t.Run(tt.name, func(t *testing.T) {
+			for validRule, expectedCost := range tt.expectCost {
+				t.Run(validRule, func(t *testing.T) {
+					s := withRule(*tt.schema, validRule)
+					celValidator := NewValidator(&s, PerCallLimit)
+					if celValidator == nil {
+						t.Fatal("expected non nil validator")
+					}
+					errs, rtCost := celValidator.getRuntimeCost(field.NewPath("root"), &s, tt.obj, 0)
+					if rtCost != expectedCost {
+						t.Fatalf("runtime cost %d does not match expected runtime cost %d", rtCost, expectedCost)
+					}
+					for _, err := range errs {
+						t.Fatalf("got unexpected error: %v", err)
+					}
+				})
+			}
+		})
+	}
+}
+
+// getRuntimeCost is to calculate runtime cost for cel validation
+func (s *Validator) getRuntimeCost(fldPath *field.Path, sts *schema.Structural, obj interface{}, existingCost uint64) (errs field.ErrorList, rtCost uint64) {
```

Comment:
it's surprising to see methods on a non-test type declared inside _test files. if we're not going to use these outside test code, making these functions that take an `s *Validator` as an arg seems more correct

Reply:
also, these traversals seem like they're duplicating a lot of the non-test remaining cost budget code ... is there a way to use the existing non-test code instead of rewriting compilation/evaluation/cost-aggregation/traversal? something like calling `Validator#Validate`, passing in MaxInt as the costBudget, asserting the returned remainingBudget is >= 0 and calculating the runtime cost with `MaxInt - remainingBudget`?

Reply:
Updated